### PR TITLE
feat(firebase_analytics): retrieves `appInstanceId` property on native platforms if available

### PR DESCRIPTION
**Pull request by willbryant**
Add an appInstanceId property on FirebaseAnalytics, exporting the API from the underlying native Firebase library implementations.

## Description

This value is needed to send server-side events through the Measurement Protocol as a part of the same session, and also useful for debugging to pair up with the data visible in DebugView and BigQuery.

## Related Issues

*Replace this paragraph with a list of issues related to this PR from the [issue database](https://github.com/flutter/flutter/issues). Indicate, which of these issues are resolved or fixed by this PR. Note that you'll have to prefix the issue numbers with flutter/flutter#.*

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`).
This will ensure a smooth and quick review process. Updating the `pubspec.yaml` and changelogs is not required.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`melos run analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change.
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/FirebaseExtended/flutterfire/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

**Review by russellwheatley — CHANGES_REQUESTED** [review]
Hey @willbryant, nice work! Couple of things I'd like to see, please:

1. Could you add another mock test for the platform interface [here](packages/firebase_analytics/firebase_analytics_platform_interface/test/mock.dart)?
2. Do you mind updating the java implementation to use the [TaskCompletionSource](https://github.com/firebase/flutterfire/pull/8583/files#diff-3b0d151d53818c29df4db13f383f4b7291d7bdaf3f31425125e4c2c09c9a4a95R152-R171) like 
 the example linked?

Thanks! 😄

**Comment by willbryant** [comment]
Thanks for the code review @russellwheatley!

I'll wait till your PR is merged and rebase, so I don't create a merge conflict.

Just to understand the second point, do we still need to use a TaskCompletionSource when we already have a Task? I think I'd need to Tasks.await to get the value out of that Task to pass to taskCompletionSource.setResult?

**Comment by russellwheatley** [comment]
> Thanks for the code review @russellwheatley!
> 
> I'll wait till your PR is merged and rebase, so I don't create a merge conflict.
> 
> Just to understand the second point, do we still need to use a TaskCompletionSource when we already have a Task? I think I'd need to Tasks.await to get the value out of that Task to pass to taskCompletionSource.setResult?

@willbryant, yes, for consistency and to use the `cachedThreadPool` . You can do the following:

```java
taskCompletionSource.setResult(Tasks.await(analytics.getAppInstanceId()));
```

**Comment by willbryant** [comment]
Cool, that's basically what I did.

I added a try/finally for consistency with the others.

File: packages/firebase_analytics/firebase_analytics/android/src/main/java/io/flutter/plugins/firebase/analytics/FlutterFirebaseAnalyticsPlugin.java [hunk 1]
```diff
@@ -256,6 +260,21 @@ private Task<Void> setDefaultEventParameters(final Map<String, Object> arguments
...
 
+  private Task<String> handleGetAppInstanceId() {
+    TaskCompletionSource<String> taskCompletionSource = new TaskCompletionSource<>();
+
+    cachedThreadPool.execute(
+        () -> {
+          try {
+            taskCompletionSource.setResult(Tasks.await(analytics.getAppInstanceId()));
+          } catch (Exception e) {
+            taskCompletionSource.setResult(null);
```
Comment (russellwheatley) [inline packages/firebase_analytics/firebase_analytics/android/src/main/java/io/flutter/plugins/firebase/analytics/FlutterFirebaseAnalyticsPlugin.java:271]:
```suggestion
            taskCompletionSource.setException(e);
```

**Review by russellwheatley — CHANGES_REQUESTED** [review]

**Review by russellwheatley — CHANGES_REQUESTED** [review]
Hey @willbryant, do you mind also updating analytics web so that it throws [this exception](https://github.com/firebase/flutterfire/blob/master/packages/firebase_database/firebase_database_web/lib/firebase_database_web.dart#L81), please?

**Comment by willbryant** [comment]
Done, and squashed.

**Comment by russellwheatley** [comment]
Hey @willbryant, do you mind also adding an [e2e test](https://github.com/firebase/flutterfire/blob/master/tests/test_driver/firebase_analytics/firebase_analytics_e2e.dart) to analytics if possible? 😄

**Comment by willbryant** [comment]
@russellwheatley I can't get e2e tests to run for the life of me, but I've pushed a draft, if you wouldn't mind approving a CI run.

**Comment by willbryant** [comment]
Struggling a bit without CI access, can't get the main e2e test directory to run locally. But, I discovered there's a sort of second copy under example/ in the plugin itself, which showed the same issue, so using that I made a bit more progress.

Hopefully the next run works better.

**Comment by russellwheatley** [comment]
@willbryant run `melos run format` and all should be well. Tip for future: comment out all e2e tests [here](https://github.com/firebase/flutterfire/blob/ea03c6b349ccd63317f96e37c85daf7e11f2a28a/tests/test_driver/driver_e2e.dart#L32-L44) and run `melos run test:e2e` for testing analytics e2e (provided you've installed `melos`).

Some of the e2e's need the emulators running to work (e.g. firebase_database) so maybe that was your problem.

I'm also presuming you've ran `melos boostrap` in the project directory 😄

**Comment by russellwheatley** [comment]
ahh, you'll have to run `flutter format .`in `tests/` directory. Melos doesn't pick that up at the moment 😞

**Comment by willbryant** [comment]
Hopefully that's it, thanks for you help :).

**Review by Lyokone — APPROVED** [review]

**Pull request merged**